docs/architecture/grpc-for-wcf-developers/rpc-types.md

```diff
@@ -0,0 +1,268 @@
+---
+title: Types of RPC - gRPC for WCF Developers
+description: A review of the types of Remote Procedure Call supported by WCF and their equivalents in gRPC
+author: markrendle
+ms.date: 09/02/2019
+---
+
+# Types of RPC
+
+As a Windows Communication Foundation (WCF) developer, you're probably used to dealing with the following types of Remote Procedure Call (RPC):
+
+- Request/Reply
+- Duplex:
+  - One-way duplex with session
+  - Full duplex with session
+- One-way
+
+It's possible to map these RPC types fairly naturally to existing gRPC concepts and this chapter will look at each of these areas in turn. Similar examples will be explored in much greater depth in [Chapter 5](migrating-wcf-to-grpc.md).
+
+| WCF | gRPC |
+| --- | ---- |
+| Regular request/reply | Unary |
+| Duplex service with session using a client callback interface | Server streaming |
+| Full duplex service with session | Bidirectional streaming |
+| One-way operations | Client streaming |
+
+## Request/reply
+
+For simple request/reply methods that take and return small amounts of data, use the simplest gRPC pattern, the unary RPC.
+
+```protobuf
+service Things {
+    rpc Get(GetThingRequest) returns (GetThingResponse);
+}
+```
+
+```csharp
+public class ThingService : Things.ThingsBase
+{
+    public override async Task<GetThingResponse> Get(GetThingRequest request, ServerCallContext context)
+    {
+        // Get thing from database
+        return new GetThingResponse { Thing = thing };
+    }
+}
+```
+
+```csharp
+public async Task ShowThing(int thingId)
+{
+    var thing = await _thingsClient.GetAsync(new GetThingRequest { ThingId = thingId });
+    Console.WriteLine($"{thing.Name}");
+}
+```
+
+As you can see, implementing a gRPC unary RPC service method is very similar to implementing a WCF operation, except that with gRPC you override a base class method instead of implementing an interface. Note that on the server, gRPC base methods always return a <xref:System.Threading.Tasks.Task%601>, although the client provides both async and blocking methods to call the service.
+
+## WCF duplex, one-way to client
+
+WCF applications (with certain bindings) can create a persistent connection between client and server, and the server can asynchronously send data to the client until the connection is closed, using a *callback interface*.
```

Comment:
Maybe be a bit more concrete and say that they do this by setting the callback service contract interface type on ServiceContract.CallbackContract.